providers/base/provider.go

```diff
@@ -90,15 +97,66 @@ func (p *Provider[K, V]) Start(ctx context.Context) error {
 		p.logger.Warn("no ids to fetch")
 	}
 
+	// If the config updater is set, the provider may update it's internal configurations
+	// on the fly. As such, we need to listen for updates to the config updater and restart
+	// the provider's main loop when the configuration changes.
+	go p.listenConfigUpdater(ctx)
+
 	// Start the main loop.
-	return p.fetch(ctx)
+	for {
+		// Create a new context for the fetch loop. This allows us to cancel the fetch loop
+		// when the provider needs to be restarted.
+		fetchCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		// Start the fetch loop.
+		errCh := make(chan error)
+		go func() {
+			errCh <- p.fetch(fetchCtx)
+		}()
+
+		select {
+		case <-p.restartCh:
+			// If any of the provider's configurations have changed, the provider will
+			// be signalled to restart.
+			p.logger.Info("restarting provider")
+			cancel()
+
+			// Wait for the fetch loop to stop.
+			err := <-errCh
+			p.logger.Debug("provider fetch loop stopped", zap.Error(err))
+		case err := <-errCh:
+			return err
+		case <-ctx.Done():
+			return <-errCh
+		}
```

Comment:
Why do we need to kill the providers if there is a config update? Don't we j need to update the set of IDs it's fetching for?

Reply:
It works cleanly for that with REST providers, but for websockets, things get more complicated.  We would need to do unsubscribe/subscribe logic based on the diff we get in providers. 

This just simplifies things, creating a new connection so we have a fresh set of subscriptions.